experimental/cuda2/graph_command_buffer.c

```diff
@@ -26,6 +25,154 @@
 // descriptor bindings and push constants.
 #define IREE_HAL_CUDA_MAX_KERNEL_ARG 128
 
+// Segmented submission management
+//
+// In a CUDA graph, buffer management and kernel launches are represented as
+// graph nodes. Dependencies are represented by graph edges. IREE's HAL follows
+// the Vulkan command buffer recording model, which "linearizes" the original
+// graph. So we have a mismatch here. Implementing IREE's HAL using CUDA graph
+// would require rediscover the graph node dependencies from the linear chain of
+// command buffer commands; it means looking at both previous and next commands
+// sometimes.
+//
+// Due to these reasons, it's beneficial to have a complete view of the full
+// command buffer and extra flexibility during recording, in order to fixup past
+// commands, or inspect past/future commands.
+//
+// Therefore, to implement IREE HAL command buffers using CUDA graphs, we
+// perform two steps using a linked list of command segments. First we create
+// segments (iree_hal_cuda2_command_buffer_prepare_*) to keep track of all IREE
+// HAL commands and the associated data, and then, when finalizing the command
+// buffer, we iterate through all the segments and record their contents
+// (iree_hal_cuda2_command_segment_record_*) into a proper CUDA graph command
+// buffer. A linked list gives us the flexibility to organize command sequence
+// in low overhead; and a deferred recording gives us the complete picture of
+// the command buffer when really started recording.
+
+//===----------------------------------------------------------------------===//
+// Command segment
+//===----------------------------------------------------------------------===//
+
+// Command action kind of a command segment.
+typedef enum iree_hal_cuda2_command_segment_action_e {
+  IREE_HAL_cuda2_COMMAND_SEGMENT_ACTION_BARRIER,
+  IREE_HAL_cuda2_COMMAND_SEGMENT_ACTION_DISPATCH,
+  IREE_HAL_cuda2_COMMAND_SEGMENT_ACTION_FILL_BUFFER,
+  IREE_HAL_cuda2_COMMAND_SEGMENT_ACTION_COPY_BUFFER,
+  IREE_HAL_cuda2_COMMAND_SEGMENT_ACTION_UPDATE_BUFFER,
+} iree_hal_cuda2_command_segment_action_t;
```

Comment:
Uppercase the full enum names?

```suggestion
// Command action kind of a command segment.
typedef enum iree_hal_cuda2_command_segment_action_e {
  IREE_HAL_CUDA2_COMMAND_SEGMENT_ACTION_BARRIER,
  IREE_HAL_CUDA2_COMMAND_SEGMENT_ACTION_DISPATCH,
  IREE_HAL_CUDA2_COMMAND_SEGMENT_ACTION_FILL_BUFFER,
  IREE_HAL_CUDA2_COMMAND_SEGMENT_ACTION_COPY_BUFFER,
  IREE_HAL_CUDA2_COMMAND_SEGMENT_ACTION_UPDATE_BUFFER,
} iree_hal_cuda2_command_segment_action_t;
```

Reply:
Oops, bad search and replace. :)